crates/runtime/src/instance/allocator/pooling/index_allocator.rs

```diff
@@ -248,59 +156,46 @@ impl IndexAllocator {
-                    // from some other module's freelist to be proportional to
-                    // that module's freelist length. Or in other words, every
-                    // *slot* should be equally likely to be stolen. The
-                    // alternative, where we pick the victim module freelist
-                    // first, means that either a module with an affinity
-                    // freelist of one slot has the same chances of losing that
-                    // slot as one with a hundred slots; or else we need a
-                    // weighted random choice among modules, which is just as
-                    // complex as this process.
-                    //
-                    // We don't bother picking an empty slot (no established
-                    // affinity) before a random slot, because this is more
-                    // complex, and in the steady state, all slots will see at
-                    // least one instantiation very quickly, so there will never
-                    // (past an initial phase) be a slot with no affinity.
-                    PoolingAllocationStrategy::ReuseAffinity => inner
-                        .pick_affine(module_id)
-                        .or_else(|| inner.pick_random())?,
+        // As a first-pass always attempt an affine allocation. This will
+        // succeed if any slots are considered affine to `module_id` (if it's
+        // specified). Failing that something else is attempted to be chosen.
+        let slot_id = inner.pick_affine(module_id).or_else(|| {
+            match mode {
+                // If any slot is requested then this is a normal instantiation
+                // looking for an index. Without any affine candidates there are
+                // two options here:
+                //
+                // 1. Pick a slot amongst previously allocated slots
+                // 2. Pick a slot that's never been used before
+                //
+                // The choice here is guided by the initial configuration of
+                // `max_unused_warm_slots`. If our unused warm slots, which are
+                // likely all affine, is below this threshold then the affinity
+                // of the warm slots isn't tampered with and first a cold slot
+                // is chosen. If the cold slot allocation fails, however, a warm
+                // slot is evicted.
+                //
+                // The opposite happens when we're above our threshold for the
+                // maximum number of warm slots, meaning that a warm slot is
+                // attempted to be picked from first with a cold slot following
+                // that.
+                AllocMode::AnySlot => {
+                    if inner.unused_warm_slots < inner.max_unused_warm_slots {
+                        inner.pick_cold().or_else(|| inner.pick_warm())
+                    } else {
+                        inner.pick_warm().or_else(|| inner.pick_cold())
```

Comment:
Similarly, I had initially thought here that `pick_warm` should always succeed so the `pick_cold` should probably be omitted in lieu of a panic (or at least a debug_assert); but this is *not* true in the case that `max_unused_warm_slots` is 0 and we're starting out with `unused_warm_slots` at 0 too. Because `<` rather than `<=` (see ablove) we'll come here even though we have no warm slots. So this is correct as written, but we should comment why.